packages/react-storage/src/components/StorageBrowser/actions/handlers/__tests__/upload.spec.ts

```diff
@@ -1,7 +1,223 @@
-import { uploadHandler } from '../upload';
+import * as StorageModule from 'aws-amplify/storage';
+
+import {
+  MULTIPART_UPLOAD_THRESHOLD_BYTES,
+  uploadHandler,
+  UploadHandlerInput,
+  UNDEFINED_CALLBACKS,
+} from '../upload';
+
+const isCancelErrorSpy = jest.spyOn(StorageModule, 'isCancelError');
+const uploadDataSpy = jest.spyOn(StorageModule, 'uploadData');
+
+const credentials = jest.fn();
+
+const config: UploadHandlerInput['config'] = {
+  accountId: 'accountId',
+  bucket: 'bucket',
+  credentials,
+  region: 'region',
+};
+
+const payload = new File([], 'test-o');
+const prefix = 'prefix';
+
+const onCancel = jest.fn();
+const onComplete = jest.fn();
+const onError = jest.fn();
+const onProgress = jest.fn();
+
+const baseInput: UploadHandlerInput = {
+  config,
+  data: { key: payload.name, payload },
+  prefix,
+};
+
+const cancel = jest.fn();
+const pause = jest.fn();
+const resume = jest.fn();
+
+const error = new Error('Failed!');
 
 describe('uploadHandler', () => {
-  it('is null', () => {
-    expect(uploadHandler).toBeNull();
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('behaves as expected in the happy path', async () => {
+    uploadDataSpy.mockReturnValueOnce({
+      cancel,
+      pause,
+      resume,
+      result: Promise.resolve({ key: payload.name }),
+      state: 'SUCCESS',
+    });
+
+    const { key, result } = uploadHandler({
+      ...baseInput,
+      options: { onComplete },
+    });
+
+    expect(await result).toBe('COMPLETE');
+
+    expect(key).toBe(baseInput.data.key);
+    expect(onComplete).toHaveBeenCalledTimes(1);
+    expect(onComplete).toHaveBeenCalledWith(baseInput.data.key);
+  });
+
+  it('calls upload with the expected values', () => {
+    uploadHandler({ ...baseInput, options: { preventOverwrite: true } });
+
+    const expected: StorageModule.UploadDataWithPathInput = {
+      data: payload,
+      options: {
+        bucket: {
+          bucketName: config.bucket,
+          region: config.region,
+        },
+        locationCredentialsProvider: credentials,
+        onProgress: expect.any(Function),
+        preventOverwrite: true,
+      },
+      path: `${prefix}${payload.name}`,
+    };
+
+    expect(uploadDataSpy).toHaveBeenCalledWith(expected);
+  });
+
+  it('calls provided onProgress callback as expected in the happy path', async () => {
+    uploadDataSpy.mockImplementation(({ options }) => {
+      // @ts-expect-error - `options` is potentially `undefined` in the `uploadData` input interface
+      options.onProgress({ totalBytes: 23, transferredBytes: 23 });
```

Comment:
nit: would this work with out the ts comment?

```suggestion
    uploadDataSpy.mockImplementation(({ options }) => {
      options?.onProgress?.({ totalBytes: 23, transferredBytes: 23 });
```